File: ptz_cameras.py
import argparse
import json
import re
import socket
import subprocess
import sys
import time
import xmlrpc.client
from threading import Thread


CAMERA_PORT = 1259
PAN_SPEED_MAX = "18"
TILT_SPEED_MAX = "14"
ON_COMMAND = "8101040002FF"
OFF_COMMAND = "8101040003FF"
PAN_TILT_ABSOLUTE_TYPE = "02"
PAN_TILT_RELATIVE_TYPE = "03"
PAN_TILT_COMMAND = "810106{PAN_TILT_TYPE}{PAN_SPEED}{TILT_SPEED}0{PAN[0]}0{PAN[1]}0{PAN[2]}0{PAN[3]}0{TILT[0]}0{TILT[1]}0{TILT[2]}0{TILT[3]}FF"
PAN_TILT_INQ_COMMAND = "81090612FF"
#[Move left 01 or right 02 or stay 03][Move Up 01 or down 02 or stay 03]
LEFT = "01"
UP = "01"
RIGHT = "02"
DOWN = "02"
STAY = "03"
PAN_TILT_DIR_COMMAND = "81010601{PAN_SPEED}{TILT_SPEED}{DIRECTION_X}{DIRECTION_Y}FF"
ZOOM_COMMAND = "810104470{ZOOM[0]}0{ZOOM[1]}0{ZOOM[2]}0{ZOOM[3]}FF"
ZOOM_INQ_COMMAND = "81090447FF"
FOCUS_COMMAND = "810104480{FOCUS[0]}0{FOCUS[1]}0{FOCUS[2]}0{FOCUS[3]}FF"
FOCUS_INQ_COMMAND = "81090448FF"
FOCUS_TYPE_AUTO="02"
FOCUS_TYPE_MANUAL="03"
FOCUS_TYPE_COMMAND = "81010438{FOCUS_TYPE}FF"
ONE_SECOND=1

def launch_daemon(debug=False):
    command = "{} data_daemon.py".format(sys.executable)
    if debug:
        command = command + " --debug"
    else:
        command = command.replace("python.exe", "pythonw.exe")
    subprocess.Popen(command, creationflags=subprocess.DETACHED_PROCESS, close_fds=True, shell=True)

def get_connection(ip):
    s = socket.socket(socket.AF_INET, socket.SOCK_DGRAM, 0)
    s.connect((ip, CAMERA_PORT))
    s.settimeout(ONE_SECOND)
    return s

def send_pan_tilt_zoom_focus(data_server, camera, conn, preset, pan_tilt_type=PAN_TILT_ABSOLUTE_TYPE):
    pan_tilt_command = bytes.fromhex(PAN_TILT_COMMAND.format(PAN_TILT_TYPE=pan_tilt_type, PAN_SPEED=PAN_SPEED_MAX, TILT_SPEED=TILT_SPEED_MAX, PAN=preset["pan"], TILT=preset["tilt"]).upper())
    conn.send(pan_tilt_command)
    #response = conn.recv(1024).hex()
    #print(response)
    zoom_command = bytes.fromhex(ZOOM_COMMAND.format(ZOOM=preset["zoom"]).upper())
    conn.send(zoom_command)
    #response = conn.recv(1024).hex()
    #print(response)
    focus_command = bytes.fromhex(FOCUS_COMMAND.format(FOCUS=preset["focus"]).upper())
    conn.send(focus_command)
    #response = conn.recv(1024).hex()
    #print(response)
    # data_server.set(camera, "current_pos", preset)

def send_commands(data_server, args, camera):
    try:
        ip = data_server.get(camera, "ip")
    except:
        launch_daemon(args.debug)
        ip = data_server.get(camera, "ip")
        print(data_server.ptz_data_size())
    conn = get_connection(ip)
    query_results = {}
    if args.off:
        preset = data_server.get(camera, "goodnight")
        if preset:
            send_pan_tilt_zoom_focus(data_server, camera, conn, preset)
            time.sleep(10)
        else:
            print("Can't find that preset")
        data = bytes.fromhex(OFF_COMMAND)
        conn.send(data)
        # response = conn.recv(1024).hex()
        # print(response)
    if args.on:
        data = bytes.fromhex(ON_COMMAND)
        conn.send(data)
    if args.preset:
        preset = data_server.get(camera, args.preset)
        if preset:
            send_pan_tilt_zoom_focus(data_server, camera, conn, preset)
        else:
            print("Can't find that preset")
    if args.pan or args.tilt or args.zoom:
        data = {
            "zoom": args.zoom if args.zoom else "0000",
            "pan": args.pan if args.pan else "0000",
            "tilt": args.tilt if args.tilt else "0000",
            "focus": "0000"
        }
        # TODO handle if data if input is bad
        send_pan_tilt_zoom_focus(data_server, camera, conn, data, PAN_TILT_RELATIVE_TYPE)
    if args.query_zoom or args.query_all:
        conn.send(bytes.fromhex(ZOOM_INQ_COMMAND))
        response = conn.recv(1024).hex().upper()
        match = re.match(r"90500(?P<p>.)0(?P<q>.)0(?P<r>.)0(?P<s>.)FF", response)
        if match:
            zoom_response = match.groupdict()
            zoom_value = "{}{}{}{}".format(zoom_response["p"], zoom_response["q"], zoom_response["r"], zoom_response["s"])
            query_results["zoom"] = zoom_value
    if args.query_pan_tilt or args.query_all:
        conn.send(bytes.fromhex(PAN_TILT_INQ_COMMAND))
        response = conn.recv(1024).hex().upper()
        match = re.match(r"90500(?P<w0>.)0(?P<w1>.)0(?P<w2>.)0(?P<w3>.)0(?P<z0>.)0(?P<z1>.)0(?P<z2>.)0(?P<z3>.)FF", response)
        if match:
            pan_tilt_response = match.groupdict()
            pan_value = "{}{}{}{}".format(pan_tilt_response["w0"], pan_tilt_response["w1"], pan_tilt_response["w2"], pan_tilt_response["w3"])
            tilt_value = "{}{}{}{}".format(pan_tilt_response["z0"], pan_tilt_response["z1"], pan_tilt_response["z2"], pan_tilt_response["z3"])
            query_results["pan"] = pan_value
            query_results["tilt"] = tilt_value
    if args.query_focus or args.query_all:
        conn.send(bytes.fromhex(FOCUS_INQ_COMMAND))
        response = conn.recv(1024).hex().upper()
        match = re.match(r"90500(?P<p>.)0(?P<q>.)0(?P<r>.)0(?P<s>.)FF", response)
        if match:
            focus_response = match.groupdict()
            focus_value = "{}{}{}{}".format(focus_response["p"], focus_response["q"], focus_response["r"], focus_response["s"])
            query_results["focus"] = focus_value
    if args.query_zoom or args.query_pan_tilt or args.query_focus or args.query_all:
        print(camera, json.dumps(query_results))
        if args.log:
            data_server.log(xmlrpc.client.dumps(({camera: query_results},)))
    conn.close()

def process_input(args):
    # Set up daemon url
    data_server = xmlrpc.client.ServerProxy('http://127.0.0.1:8001')
    # Launch daemon
    if args.launch_daemon:
        launch_daemon(args.debug)
    
    if args.main_action:
        camera = "main"
        send_commands(data_server, args, camera)

    if args.alt_action:
        camera = "alt"
        send_commands(data_server, args, camera)

    # threads = []
    # # Create Main camera thread
    # if args.main_action:
    #     camera = "main"
    #     threads.append(Thread(target=send_commands, args=(data_server, args, camera)))

    # # Create Alt camera thread
    # if args.alt_action:
    #     camera = "alt"
    #     threads.append(Thread(target=send_commands, args=(data_server, args, camera)))

    # # Execute the threads
    # for thread in threads:
    #     thread.start()

    # # Join the completed threads
    # for thread in threads:
    #     thread.join()

    # Shutdown daemon
    if args.stop_daemon or args.off:
        data_server.stop_daemon()


if __name__ == "__main__":
    parser = argparse.ArgumentParser(description='Interact with PTZ cameras')
    parser.add_argument('--launch_daemon', dest='launch_daemon', action='store_true', default=False, help='Start the Daemon (This should be done automatically)')
    parser.add_argument('--stop_daemon', dest='stop_daemon', action='store_true', default=False, help='Stop the Daemon (This should occur automatically when cameras are shut off)')
    parser.add_argument('--main', dest='main_action', action='store_true', default=False, help='Do these actions to the main camera')
    parser.add_argument('--alt', dest='alt_action', action='store_true', default=False, help='Do these actions to the alt camera')
    parser.add_argument('--preset', dest='preset', type=str, help='Move Camera to position --preset [value]')
    parser.add_argument('--pan', dest='pan', type=str, help='Set the relative pan --pan [value]')
    parser.add_argument('--tilt', dest='tilt', type=str, help='Set the relative tilt --tilt [value]')
    parser.add_argument('--zoom', dest='zoom', type=str, help='Set the relative zoom --zoom [value]')
    parser.add_argument('--off', dest='off', action='store_true', default=False, help='Send off to camera')
    parser.add_argument('--on', dest='on', action='store_true', default=False, help='Send on to camera')
    parser.add_argument('--query_zoom', dest='query_zoom', action='store_true', default=False, help='Get the current zoom values')
    parser.add_argument('--query_pan_tilt', dest='query_pan_tilt', action='store_true', default=False, help='Get the current pan and tilt values')
    parser.add_argument('--query_focus', dest='query_focus', action='store_true', default=False, help='Get the current focus values')
    parser.add_argument('--query_all', dest='query_all', action='store_true', default=False, help='Run all Queries')
    parser.add_argument('--log', dest='log', action='store_true', default=False, help='Add to ptz_data')
    parser.add_argument('--debug', dest='debug', action='store_true', default=False, help='Launch debug windows')
    
    args = parser.parse_args()
    process_input(args)

File: ptz_cameras.js
// Needed for socket communication
const net = require('net');
// Needed for the closest thing I can find to fstring behavior
const util = require('util');
// Needed to read a file
const fs = require('fs');
// Obtain the arguments and pass to minimist to put the args in a map
// npm install minimist
const args = require('minimist')(process.argv);

// Help then exit 
if ("help" in args) {
    console.log('\nnode ptz_cameras.js [commands]\n');
    console.log('\t--main', 'Do these actions to the main camera');
    console.log('\t--alt', 'Do these actions to the alt camera');
    console.log('\t--preset', 'Move Camera to position --preset [value]');
    console.log('\t--pan', 'Set the relative pan --pan [value]');
    console.log('\t--tilt', 'Set the relative tilt --tilt [value]');
    console.log('\t--zoom', 'Set the relative zoom --zoom [value]');
    console.log('\t--off', 'Send off to camera');
    console.log('\t--on', 'Send on to camera');
    console.log('\t--query_zoom', 'Get the current zoom values');
    console.log('\t--query_pan_tilt', 'Get the current pan and tilt values');
    console.log('\t--query_focus', 'Get the current focus values');
    console.log('\t--query_all', 'Run all Queries');
    process.exit(0);
}

// Define constants
const CAMERA_PORT = 1259;
const PAN_SPEED_MAX = "18";
const TILT_SPEED_MAX = "14";
const ON_COMMAND = "8101040002FF";
const OFF_COMMAND = "8101040003FF";
const PAN_TILT_ABSOLUTE_TYPE = "02";
const PAN_TILT_RELATIVE_TYPE = "03";
const PAN_TILT_COMMAND = "810106%s%s%s0%s0%s0%s0%s0%s0%s0%s0%sFF";
const PAN_TILT_INQ_COMMAND = "81090612FF";
// [Move left 01 or right 02 or stay 03][Move Up 01 or down 02 or stay 03]
const LEFT = "01";
const UP = "01";
const RIGHT = "02";
const DOWN = "02";
const STAY = "03";
const PAN_TILT_DIR_COMMAND = "81010601%s%s%s%sFF";
const ZOOM_COMMAND = "810104470%s0%s0%s0%sFF";
const ZOOM_INQ_COMMAND = "81090447FF";
const FOCUS_COMMAND = "810104480%s0%s0%s0%sFF";
const FOCUS_INQ_COMMAND = "81090448FF";
const FOCUS_TYPE_AUTO = "02";
const FOCUS_TYPE_MANUAL = "03";
const FOCUS_TYPE_COMMAND = "81010438%sFF";
const ONE_SECOND_IN_MS = 1000;
const TEN_SECONDS_IN_MS = 10000;


function sleep(ms) {
    return new Promise((resolve) => {
        setTimeout(resolve, ms);
    });
}

function send_command(socket, ip, command) {
    return new Promise(function (resolve, reject) {
        let command_hex = Buffer.from(command, 'hex');
        socket.send(command_hex, 0, Buffer.byteLength(command_hex), CAMERA_PORT, ip);
        socket.on("message", data => {
            resolve(data);
        });
    });
}


async function send_pan_tilt_zoom_focus(camera, socket, ip, preset, pan_tilt_type = PAN_TILT_ABSOLUTE_TYPE) {
    const pan_tilt_command = util.format(PAN_TILT_COMMAND, pan_tilt_type, PAN_SPEED_MAX, TILT_SPEED_MAX, preset.pan[0], preset.pan[1], preset.pan[2], preset.pan[3], preset.tilt[0], preset.tilt[1], preset.tilt[2], preset.tilt[3]).toUpperCase();
    const pan_tilt_command_hex = Buffer.from(pan_tilt_command, 'hex');
    await send_command(socket, ip, pan_tilt_command);

    const zoom_command = util.format(ZOOM_COMMAND, preset.zoom[0], preset.zoom[1], preset.zoom[2], preset.zoom[3]).toUpperCase();
    const zoom_command_hex = Buffer.from(zoom_command, 'hex');
    await send_command(socket, ip, zoom_command);

    const focus_command = util.format(FOCUS_COMMAND, preset.focus[0], preset.focus[1], preset.focus[2], preset.focus[3]).toUpperCase();
    const focus_command_hex = Buffer.from(focus_command, 'hex');
    await send_command(socket, ip, focus_command);
}

function query_zoom(socket, ip) {
    return new Promise(function (resolve, reject) {
        let zoom_inq_command_hex = Buffer.from(ZOOM_INQ_COMMAND, 'hex');
        socket.send(zoom_inq_command_hex, 0, Buffer.byteLength(zoom_inq_command_hex), CAMERA_PORT, ip);
        socket.on("message", data => {
            const response = data.toString('hex').toUpperCase();
            const regExZoomQuery = /90500(?<p>.)0(?<q>.)0(?<r>.)0(?<s>.)FF/mg;
            for (const match of response.matchAll(regExZoomQuery)) {
                if (match.groups !== undefined) {
                    let zoom_value = `${match.groups.p}${match.groups.q}${match.groups.r}${match.groups.s}`;
                    resolve(zoom_value);
                } else {
                    reject("No Matches");
                }
            }
        });
    });
}

function query_pan_tilt(socket, ip) {
    return new Promise(function (resolve, reject) {
        let pan_tilt_inq_command_hex = Buffer.from(PAN_TILT_INQ_COMMAND, 'hex');
        socket.send(pan_tilt_inq_command_hex, 0, Buffer.byteLength(pan_tilt_inq_command_hex), CAMERA_PORT, ip);
        socket.on("message", data => {
            const response = data.toString('hex').toUpperCase();
            const regExPanTiltQuery = /90500(?<w0>.)0(?<w1>.)0(?<w2>.)0(?<w3>.)0(?<z0>.)0(?<z1>.)0(?<z2>.)0(?<z3>.)FF/mg;
            for (const match of response.matchAll(regExPanTiltQuery)) {
                if (match.groups !== undefined) {
                    let pan_value = `${match.groups.w0}${match.groups.w1}${match.groups.w2}${match.groups.w3}`;
                    let tilt_value = `${match.groups.z0}${match.groups.z1}${match.groups.z2}${match.groups.z3}`;
                    resolve([pan_value, tilt_value]);
                } else {
                    reject("No Matches");
                }
            }
        });
    });
}

function query_focus(socket, ip) {
    return new Promise(function (resolve, reject) {
        let focus_inq_command_hex = Buffer.from(FOCUS_INQ_COMMAND, 'hex');
        socket.send(focus_inq_command_hex, 0, Buffer.byteLength(focus_inq_command_hex), CAMERA_PORT, ip);
        socket.on("message", data => {
            const response = data.toString('hex').toUpperCase();
            const regExFocusQuery = /90500(?<p>.)0(?<q>.)0(?<r>.)0(?<s>.)FF/mg;
            for (const match of response.matchAll(regExFocusQuery)) {
                if (match.groups !== undefined) {
                    let focus_value = `${match.groups.p}${match.groups.q}${match.groups.r}${match.groups.s}`;
                    resolve(focus_value);
                } else {
                    reject("No Matches");
                }
            }
        });
    });
}

async function send_commands(ptz_data, camera) {
    let ip = ptz_data[camera]["ip"];
    
    let socket = require('dgram').createSocket('udp4');
    if ("off" in args) {
        // Sends the off command for the camera
        let preset = ptz_data[camera]["goodnight"];
        if (preset !== undefined) {
            await send_pan_tilt_zoom_focus(camera, socket, ip, preset);
            await sleep(TEN_SECONDS_IN_MS);
        } else {
            console.log("Can't find that preset");
        }
        await send_command(socket, ip, OFF_COMMAND);
    } else if ("on" in args) {
        // Sends the on command for the camera
        await send_command(socket, ip, ON_COMMAND);
    } else if ("preset" in args) {
        // Sends whatever preset data is available for the camera
        let preset = ptz_data[camera][args.preset];
        if (preset !== undefined) {
            await send_pan_tilt_zoom_focus(camera, socket, ip, preset);
        } else {
            console.log("Can't find that preset");
        }
    } else if ("pan" in args || "tilt" in args || "zoom" in args) {
        // Sends a relative pan tilt zoom to the camera
        let data = {
            "zoom": (zoom in args) ? args.zoom : "0000",
            "pan": (pan in args) ? args.pan : "0000",
            "tilt": (tilt in args) ? args.tilt : "0000",
            "focus": "0000"
        };
        await send_pan_tilt_zoom_focus(camera, socket, ip, data, PAN_TILT_RELATIVE_TYPE);
    }

    let query_results = {};
    if ("query_zoom" in args || "query_all" in args) {
        query_results["zoom"] = await query_zoom(socket, ip);
    }
    if ("query_pan_tilt" in args || "query_all" in args) {
        [query_results["pan"], query_results["tilt"]] = await query_pan_tilt(socket, ip);
    }
    if ("query_focus" in args || "query_all" in args) {
        query_results["focus"] = await query_focus(socket, ip);
    }
    if ("query_zoom" in args || "query_pan_tilt" in args || "query_focus" in args || "query_all" in args) {
        console.log(camera, query_results);
    }
    socket.close();
}

let ptzRawData = fs.readFileSync('ptz_data.json');
let ptz_data = JSON.parse(ptzRawData);
if ("main" in args) {
    send_commands(ptz_data, "main");
}

if ("alt" in args) {
    send_commands(ptz_data, "alt");
}

